### ISSUE TITLE
Replace log4j2 with tinylog2 + slf4j bridge

### ISSUE THREAD

**Issue by Siedlerchr** [JabRef/jabref#8007]
We always get this Logger error messages on startup and we use a snapshot version of log4j2 which will probably take a long time until it's stable.

We already use slf4j as a Logger interface, so switching to another underlying logger should not be that complicated. Most stuff should be compatible.
We currently use slf4j-2.0.0. This requires at least [tinylog 2.4](https://github.com/tinylog-org/tinylog/pull/199/commits)

https://tinylog.org/v2/download/#slf4j 

## What needs to be changed?

1. Replace log4j2 with tinylog and the bindings.
2. Investigate if we still need our custom Logging Plugins or if they need to be converted  https://tinylog.org/v2/extending/ 
3. Remove log4j stuff also from jpackage/jlink 
4. Test if it works on all platforms

**Comment by koppor** [comment]
Historic remark: In 2017, I replaced our hole logging by jcabi framework at https://github.com/JabRef/jabref/pull/3015.

Two concerns:

1. The method signature for logging exceptions was very different (https://github.com/JabRef/jabref/pull/3015#issuecomment-315673569)
2. No SLF4J was used.

In last devcall we discussed that we want to keep slf4j. - For me, it would be OK to even go away from slf4j for logging our things in case this increases startup time.

**Comment by koppor** [comment]
This will probably fix #5475 (refs https://github.com/JabRef/jabref/issues/5475#issuecomment-858083303).

**Comment by ThiloteE** [comment]
log4j seems to have a critical software issue. Germany's federal ministry for information security put it on code red: 

- https://www.bsi.bund.de/EN/Home/home_node.html
- https://www.bsi.bund.de/SharedDocs/Cybersicherheitswarnungen/DE/2021/2021-549032-10F2.pdf?__blob=publicationFile&v=8

**Comment by Siedlerchr** [comment]
JabRef uses version 3.x snapshot. Gets automatically updated. And JabRef does not really log user input and is not running on a server.

**Comment by ThiloteE** [comment]
Ah nice, this critical security issue seems to affect versions 2 .0 - 2.14 only. Not relevant for Jabref then :)

**Comment by calixtus** [comment]
We expect to be able to release 5.4 in the next few days / two weeks.
But to adress your concerns, I just checked 5.3: it's using log4j-3-snapshot too.

**Comment by navoooo** [comment]
> And JabRef does not really log user input and is not running on a server.

It does not have to run on a server to be targeted, a compromised BibTeX/BibLaTeX database distributed by mail or downloaded could've been enough if JabRef was vulnerable. The file format is commonly exchanged by various means and is not generally suspected of malicious content, so if it were possible to load executable code through this with a well-placed string, it's not hard to imagine it being used as a delivery method for worms or ransomware targeting researchers.

It at least appeared to me as if there might be some candidates too, e.g. [from bibfile directly](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/exporter/BibTeXMLExporter.java#L214), [from PDF annotations](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/pdf/PdfAnnotationImporter.java#L83) and [presumably user input](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/pdf/search/retrieval/PdfSearcher.java#L72), but I did not look into it further after finding the log4j version string.

**Comment by dengelt** [comment]
> But to adress your concerns, I just checked 5.3: it's using log4j-3-snapshot too.

Does that really mean it's not vulnerable though? Isn't the "3.0.0-SNAPSHOT" version of log4j just whatever is currently on the master branch? At least I can't find any actual released versions with that version number anywhere, and nobody seems to mention it when talking about the log4shell vulnerability.

If that is the case, then anyone building JabRef before 12th December would have a vulnerable version (including the 5.3 binaries distributed on the release page that were built on 5th July).

**Comment by Siedlerchr** [comment]
No idea though, but seems possible. There might be more differences between 2.x and 3.x. However, we are preparing a release in the next days which will then include the latest available version.
Despite that, the attack surface in JabRef is extremely low. I checked if I can trigger this while importing (e.g a modified bib file et but could not trigger the vuln..

**Comment by koppor** [comment]
Just for the record:

- log4j hotpatch for running JVM instances: https://github.com/corretto/hotpatch-for-apache-log4j2
- patch for new JVM instances: https://github.com/nccgroup/log4j-jndi-be-gone

**Comment by cinderbdt** [comment]
Could you comment on where on the Windows distribution one would replace the log4j2?
`│   jabref-chrome.json
│   jabref-firefox.json
│   JabRef.exe
│   JabRef.ico
│   JabRefHost.bat
│   JabRefHost.ps1
│
├───app
│       JabRef.cfg
│
└───runtime
    │   release
    │
    ├───bin
    │   │   api-ms-win-core-console-l1-1-0.dll
    │   │   api-ms-win-core-datetime-l1-1-0.dll
    │   │   api-ms-win-core-debug-l1-1-0.dll
    │   │   api-ms-win-core-errorhandling-l1-1-0.dll
    │   │   api-ms-win-core-file-l1-1-0.dll
    │   │   api-ms-win-core-file-l1-2-0.dll
    │   │   api-ms-win-core-file-l2-1-0.dll
    │   │   api-ms-win-core-handle-l1-1-0.dll
    │   │   api-ms-win-core-heap-l1-1-0.dll
    │   │   api-ms-win-core-interlocked-l1-1-0.dll
    │   │   api-ms-win-core-libraryloader-l1-1-0.dll
    │   │   api-ms-win-core-localization-l1-2-0.dll
    │   │   api-ms-win-core-memory-l1-1-0.dll
    │   │   api-ms-win-core-namedpipe-l1-1-0.dll
    │   │   api-ms-win-core-processenvironment-l1-1-0.dll
    │   │   api-ms-win-core-processthreads-l1-1-0.dll
    │   │   api-ms-win-core-processthreads-l1-1-1.dll
    │   │   api-ms-win-core-profile-l1-1-0.dll
    │   │   api-ms-win-core-rtlsupport-l1-1-0.dll
    │   │   api-ms-win-core-string-l1-1-0.dll
    │   │   api-ms-win-core-synch-l1-1-0.dll
    │   │   api-ms-win-core-synch-l1-2-0.dll
    │   │   api-ms-win-core-sysinfo-l1-1-0.dll
    │   │   api-ms-win-core-timezone-l1-1-0.dll
    │   │   api-ms-win-core-util-l1-1-0.dll
    │   │   api-ms-win-crt-conio-l1-1-0.dll
    │   │   api-ms-win-crt-convert-l1-1-0.dll
    │   │   api-ms-win-crt-environment-l1-1-0.dll
    │   │   api-ms-win-crt-filesystem-l1-1-0.dll
    │   │   api-ms-win-crt-heap-l1-1-0.dll
    │   │   api-ms-win-crt-locale-l1-1-0.dll
    │   │   api-ms-win-crt-math-l1-1-0.dll
    │   │   api-ms-win-crt-multibyte-l1-1-0.dll
    │   │   api-ms-win-crt-private-l1-1-0.dll
    │   │   api-ms-win-crt-process-l1-1-0.dll
    │   │   api-ms-win-crt-runtime-l1-1-0.dll
    │   │   api-ms-win-crt-stdio-l1-1-0.dll
    │   │   api-ms-win-crt-string-l1-1-0.dll
    │   │   api-ms-win-crt-time-l1-1-0.dll
    │   │   api-ms-win-crt-utility-l1-1-0.dll
    │   │   awt.dll
    │   │   fontmanager.dll
    │   │   freetype.dll
    │   │   harfbuzz.dll
    │   │   j2gss.dll
    │   │   j2pcsc.dll
    │   │   j2pkcs11.dll
    │   │   jaas.dll
    │   │   JabRef
    │   │   JabRef.bat
    │   │   jabswitch.exe
    │   │   jaccessinspector.exe
    │   │   jaccesswalker.exe
    │   │   jar.exe
    │   │   jarsigner.exe
    │   │   java.dll
    │   │   java.exe
    │   │   javaaccessbridge.dll
    │   │   javac.exe
    │   │   javadoc.exe
    │   │   javajpeg.dll
    │   │   javap.exe
    │   │   javaw.exe
    │   │   jawt.dll
    │   │   jdeprscan.exe
    │   │   jdeps.exe
    │   │   jfr.exe
    │   │   jimage.dll
    │   │   jimage.exe
    │   │   jli.dll
    │   │   jlink.exe
    │   │   jmod.exe
    │   │   jpackage.dll
    │   │   jpackage.exe
    │   │   jrunscript.exe
    │   │   jsound.dll
    │   │   keytool.exe
    │   │   kinit.exe
    │   │   klist.exe
    │   │   ktab.exe
    │   │   lcms.dll
    │   │   management.dll
    │   │   management_ext.dll
    │   │   mlib_image.dll
    │   │   msvcp140.dll
    │   │   net.dll
    │   │   nio.dll
    │   │   prefs.dll
    │   │   rmi.dll
    │   │   rmid.exe
    │   │   rmiregistry.exe
    │   │   serialver.exe
    │   │   splashscreen.dll
    │   │   sspi_bridge.dll
    │   │   sunmscapi.dll
    │   │   ucrtbase.dll
    │   │   vcruntime140.dll
    │   │   vcruntime140_1.dll
    │   │   verify.dll
    │   │   w2k_lsa_auth.dll
    │   │   windowsaccessbridge-64.dll
    │   │   zip.dll
    │   │
    │   └───server
    │           jvm.dll
    │
    ├───conf
    │   │   logging.properties
    │   │   net.properties
    │   │   sound.properties
    │   │
    │   └───security
    │       │   java.policy
    │       │   java.security
    │       │
    │       └───policy
    │           │   README.txt
    │           │
    │           ├───limited
    │           │       default_local.policy
    │           │       default_US_export.policy
    │           │       exempt_local.policy
    │           │
    │           └───unlimited
    │                   default_local.policy
    │                   default_US_export.policy
    │
    ├───legal
    │   ├───java.base
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       aes.md
    │   │       asm.md
    │   │       ASSEMBLY_EXCEPTION
    │   │       c-libutl.md
    │   │       cldr.md
    │   │       icu.md
    │   │       LICENSE
    │   │       public_suffix.md
    │   │       unicode.md
    │   │       zlib.md
    │   │
    │   ├───java.compiler
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.datatransfer
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.desktop
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       colorimaging.md
    │   │       freetype.md
    │   │       giflib.md
    │   │       harfbuzz.md
    │   │       jpeg.md
    │   │       lcms.md
    │   │       libpng.md
    │   │       LICENSE
    │   │       mesa3d.md
    │   │
    │   ├───java.logging
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.management
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.management.rmi
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.naming
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.net.http
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.prefs
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.rmi
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.scripting
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.security.jgss
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.security.sasl
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.smartcardio
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.sql
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.sql.rowset
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.transaction.xa
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───java.xml
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       bcel.md
    │   │       dom.md
    │   │       jcup.md
    │   │       LICENSE
    │   │       xalan.md
    │   │       xerces.md
    │   │
    │   ├───java.xml.crypto
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │       santuario.md
    │   │
    │   ├───jdk.accessibility
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.charsets
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.compiler
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.crypto.cryptoki
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │       pkcs11cryptotoken.md
    │   │       pkcs11wrapper.md
    │   │
    │   ├───jdk.crypto.ec
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.crypto.mscapi
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.internal.opt
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       jopt-simple.md
    │   │       LICENSE
    │   │
    │   ├───jdk.jartool
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.javadoc
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       jquery.md
    │   │       jqueryUI.md
    │   │       LICENSE
    │   │
    │   ├───jdk.jdeps
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.jfr
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.jlink
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.jpackage
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.jsobject
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.localedata
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       cldr.md
    │   │       LICENSE
    │   │       thaidict.md
    │   │
    │   ├───jdk.management
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.management.jfr
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.naming.dns
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.naming.rmi
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.security.auth
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.security.jgss
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.unsupported
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.unsupported.desktop
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   ├───jdk.xml.dom
    │   │       ADDITIONAL_LICENSE_INFO
    │   │       ASSEMBLY_EXCEPTION
    │   │       LICENSE
    │   │
    │   └───jdk.zipfs
    │           ADDITIONAL_LICENSE_INFO
    │           ASSEMBLY_EXCEPTION
    │           LICENSE
    │
    └───lib
        │   classlist
        │   ct.sym
        │   fontconfig.bfc
        │   fontconfig.properties.src
        │   jawt.lib
        │   jrt-fs.jar
        │   jvm.cfg
        │   jvm.lib
        │   modules
        │   psfont.properties.ja
        │   psfontj2d.properties
        │   tzdb.dat
        │   tzmappings
        │
        ├───jfr
        │       default.jfc
        │       profile.jfc
        │
        └───security
                blacklisted.certs
                cacerts
                default.policy
                public_suffix_list.dat`

**Comment by Siedlerchr** [comment]
You would need to put the agent  into the folder runtime\bin and modify the jabref.bat start script and add it to the command line arguments before the "-m" e
For linux and mac use the JabRef shellscript starter.

```
pushd %DIR% & %JAVA_EXEC% %CDS_JVM_OPTS%  -p "%~dp0/../app" -javaagent:path/to/log4j-jndi-be-gone-1.0.0-standalone.jar -m org.jabref/org.jabref.gui.JabRefLauncher  %* & popd
```

**Comment by calixtus** [comment]
Good news, we just merged #8226 into the main branch, so in 5.4 Tinylog will be used instead of log4j.
We plan to release before christmas.
